### PR TITLE
Issue #596: Add support for closing clips.

### PR DESCRIPTION
**Pull request by Julian-O**
* Add key support for close()

   * FFMPEG_VideoWriter and FFMPEG_AudioWriter: Support close() and context managers.
   * Clip: support close() and context manager. Doesn't do anything itself. The work is done in the subclasses that need it.
   * Clip subclasses: Overrride close.
       * Move away from depending on clients calling__del__(). Deleting can be left to Garbage Collector.
   * CompositeVideoClip: Note: Don't close anything that wasn't constructed here. The client needs to be able to control the component clips.
   * AudioFileClip:  Was concerned that lambda might include a reference to reader that wasn't cleaned up by close, so changed it over to an equivalent self.reader. Probably has no effect, but feels safer.

* Update tests to use close().

   * Note: While many tests pass on Linux either way, a large proportion of the existing unit tests fail on Windows without these changes.
   * Include changes to many doctest examples - Demonstrate good practice in the examples.
   * Also, migrate tests to use TEMPDIR where they were not using it.
   * test_duration(): also corrected a bug in the test (described in #598). This bug is also been addressed in #585, so a merge will be required.

* Add two new test files:

   *  test_resourcereleasedemo exercises the path where close is not called and demonstrates that there is a consistent problem on Windows. Even after this fix, it remains a problem that if you don't call close, moviepg will leak locked files and subprocesses. Because the problem remains until the process ends, this is included in a separate test file.]
   * test_resourcerelease demonstrates that when close() is called, the problem goes away.

* Update documentation to include usage tips for close()

Not included:

*  Example code has not been updated to use close().

### CODE REVIEW

**Comment by kencochrane** [comment]
Thank you for adding this, the code looks good to me, but I'm not a contributor, so we will need a contributor to review as well.

File: moviepy/audio/io/ffmpeg_audiowriter.py [hunk 1]
```diff
@@ -135,12 +135,32 @@ def __del__(self):
...
+
+    # Support the Context Manager protocol, to ensure that resources are cleaned up.
+
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
         self.close()
 
+    def __exit__(self, exc_type, exc_value, traceback):
```
Comment (mgaitan) [inline moviepy/audio/io/ffmpeg_audiowriter.py]: this es repeated, right?

Comment (Julian-O) [reply moviepy/audio/io/ffmpeg_audiowriter.py]: You are right. Well spotted. That is an error. (It is of no consequence, but one should be deleted.)

**Comment by Gronis** [comment]
Hi,

Why is this issue closed? Is it fixed somewhere else?

**Pull request closed without merging**